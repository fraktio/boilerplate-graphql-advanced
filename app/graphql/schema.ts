import {
  IResolverValidationOptions,
  makeExecutableSchema,
  IExecutableSchemaDefinition,
} from "apollo-server-express";
import { GraphQLSchema } from "graphql";

import { createSchemaDirectives } from "~/graphql/directives/directives";
import { resolvers } from "~/graphql/resolvers/resolvers";
import { typeDefs } from "~/graphql/typeDefs";

export const createResolverValidationOptions = (
  behaviour: boolean,
): IResolverValidationOptions => ({
  requireResolversForArgs: behaviour,
  requireResolversForResolveType: behaviour,
  allowResolversNotInSchema: behaviour,
});

export const createSchema = (
  opts?: Partial<IExecutableSchemaDefinition>,
): IExecutableSchemaDefinition => ({
  allowUndefinedInResolve: false,
  inheritResolversFromInterfaces: true,
  resolverValidationOptions: createResolverValidationOptions(true),
  resolvers,
  typeDefs,
  schemaDirectives: createSchemaDirectives(),
  ...opts,
});

export const createExecutableSchema = (
  opts?: Partial<IExecutableSchemaDefinition>,
): GraphQLSchema => makeExecutableSchema(createSchema(opts));
