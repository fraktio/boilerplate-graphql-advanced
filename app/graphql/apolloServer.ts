import { ApolloServer, ApolloServerExpressConfig } from "apollo-server-express";

import { Config } from "~/config/config";
import { createDataSources } from "~/dataSources";
import { apolloErrorHandler } from "~/express/middleware/errorHandler";
import { createPlugins } from "~/graphql/plugins/plugins";
import { createExecutableSchema } from "~/graphql/schema";
import { createValidationRules } from "~/graphql/validationRules";

type CreateServerOpts = ApolloServerExpressConfig & {
  config: Config;
};

export const createApolloServer = ({
  config,
  ...rest
}: CreateServerOpts): ApolloServer =>
  new ApolloServer({
    validationRules: createValidationRules(),
    schema: createExecutableSchema(),
    formatError: apolloErrorHandler({ config }),
    plugins: createPlugins({ graphqlConfig: config.graphql }),
    introspection: config.graphql.allowIntrospection,
    dataSources: createDataSources,
    persistedQueries: { ttl: 43200 /* 12h */ },
    ...rest,
  });
