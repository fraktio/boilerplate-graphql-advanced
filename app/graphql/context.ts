import Logger from "bunyan";
import { Request, Response } from "express";

import { Config } from "~/config";
import { DataSourcesInContext } from "~/dataSources";
import { DBConnection } from "~/database/connection";
import { createDataLoaders, DataLoaders } from "~/database/dataLoaders";
import { UserTable } from "~/database/user/userDatabase";
import { Maybe } from "~/generation/generated";

export type Context = BaseContext & DataSourcesInContext;

export type BaseContext = {
  logger: Logger;
  req: Request;
  res: Response;
  config: Config;
  knex: DBConnection;
  dataLoaders: DataLoaders;
  authenticatedUser: Maybe<UserTable>;
};

export const createContext = (params: {
  knex: DBConnection;
  config: Config;
}) => (app: { req: Request; res: Response }) => ({
  logger: app.req.logger,
  req: app.req,
  res: app.res,
  config: params.config,
  knex: params.knex,
  dataLoaders: createDataLoaders(),
  authenticatedUser: app.req.user || null,
});
