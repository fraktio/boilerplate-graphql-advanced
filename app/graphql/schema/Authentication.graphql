type AuthenticatedUserSuccess {
  user: User!
}

type AuthenticatedUserFailure {
  success: Boolean!
}

union AuthenticatedUserResponse = AuthenticatedUserSuccess | AuthenticatedUserFailure

type Query {
  authenticatedUser: AuthenticatedUserResponse!
}

type LoginUserSuccess {
  token: String!
  user: User!
}

type LoginUserFailure {
  success: Boolean!
}

union LoginUserResponse = LoginUserSuccess | LoginUserFailure

input LoginUserInput {
  username: String!
  password: String!
}

type Mutation {
  login(input: LoginUserInput!): LoginUserResponse!
  logout: Boolean!
}