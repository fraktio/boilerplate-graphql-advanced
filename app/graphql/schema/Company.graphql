type Company {
  UUID: UUID!
  name: String!
  timestamp: Timestamp!
  employees: [Adult!]!
}

input CompanyQuery {
  UUID: UUID!
}

type Query {
  companies: [Company!]!
  company(input: CompanyQuery!): Company!
}

type Mutation {
  addCompany(input: AddCompanyInput!): AddCompanyOutput!
  editCompany(input: EditCompanyInput!): EditCompanyOutput!
}

input AddCompanyInput {
  company: CompanyInput!
}

input EditCompanyInput {
  UUID: UUID!
  company: CompanyInput!
}

input CompanyInput {
  name: String!
  # more later..
}

union AddCompanyOutput = AddCompanySuccess | UniqueConstraintViolationFailure

type AddCompanySuccess {
  company: Company!
}

union EditCompanyOutput = EditCompanySuccess #| possible other failures

type EditCompanySuccess {
  company: Company!
}
