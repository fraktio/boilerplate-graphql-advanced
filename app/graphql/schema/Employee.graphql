type Mutation {
  addEmployee(
    input: AddEmployeeInput!
  ): AddEmployeeOutput!
  removeEmployee(
    input: RemoveEmployeeInput!
  ): RemoveEmployeeOutput!
}

input AddEmployeeInput {
  companyUUID: UUID!
  personUUID: UUID!
}

input RemoveEmployeeInput {
  companyUUID: UUID!
  personUUID: UUID!
}

union AddEmployeeOutput = AddEmployeeSuccess #| possible other failures

type AddEmployeeSuccess {
  company: Company!
}

union RemoveEmployeeOutput = RemoveEmployeeSuccess #| possible other failures

type RemoveEmployeeSuccess {
  company: Company!
}