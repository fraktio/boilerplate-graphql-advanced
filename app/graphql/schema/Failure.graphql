interface FailureOutput {
  message: String!
  field: String!
}

"Operation fails because some value is not unique"
type UniqueConstraintViolationFailure implements FailureOutput {
  message: String!
  field: String!
}
