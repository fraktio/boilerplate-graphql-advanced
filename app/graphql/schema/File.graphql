type File {
  encoding: String!
  filename: String!
  mimetype: String!
}

type FileMetadataSuccess {
  metadata: File!
}

type FileMetadataInvalidFile implements FailureOutput {
  message: String!
  field: String!
}

union FileMetadataResponse = FileMetadataSuccess |  FileMetadataInvalidFile

type Mutation {
  fileMetadata(file: Upload!): FileMetadataResponse!
}
