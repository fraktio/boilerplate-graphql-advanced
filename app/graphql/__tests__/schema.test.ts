import { createExecutableSchema, createSchema } from "~/graphql/schema";

describe("Graphql / schema", () => {
  it("createSchema", async () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { resolvers, typeDefs, ...rest } = createSchema();

    expect(rest).toMatchSnapshot();
  });

  it("createExecutableSchema", async () => {
    const schema = createExecutableSchema();

    expect(schema).toMatchSnapshot();
  });
});
