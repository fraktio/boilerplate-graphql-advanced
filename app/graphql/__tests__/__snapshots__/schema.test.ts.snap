// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphql / schema createExecutableSchema 1`] = `
GraphQLSchema {
  "__validationErrors": undefined,
  "_directives": Array [
    "@auth",
    "@deprecated",
    "@include",
    "@skip",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {
    "FailureOutput": Object {
      "interfaces": Array [],
      "objects": Array [
        "UniqueConstraintViolationFailure",
        "NotFoundFailure",
        "InvalidCursorFailure",
      ],
    },
    "Person": Object {
      "interfaces": Array [],
      "objects": Array [
        "Adult",
        "Underage",
      ],
    },
  },
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": "Subscription",
  "_typeMap": Object {
    "AccessRight": "AccessRight",
    "AddCompanyInput": "AddCompanyInput",
    "AddCompanyOutput": "AddCompanyOutput",
    "AddCompanySuccess": "AddCompanySuccess",
    "AddEmployeeInput": "AddEmployeeInput",
    "AddEmployeeOutput": "AddEmployeeOutput",
    "AddEmployeeSuccess": "AddEmployeeSuccess",
    "AddPersonInput": "AddPersonInput",
    "AddPersonOutput": "AddPersonOutput",
    "AddPersonPersonInput": "AddPersonPersonInput",
    "AddPersonSuccess": "AddPersonSuccess",
    "Adult": "Adult",
    "AuthenticatedUserFailure": "AuthenticatedUserFailure",
    "AuthenticatedUserResponse": "AuthenticatedUserResponse",
    "AuthenticatedUserSuccess": "AuthenticatedUserSuccess",
    "Boolean": "Boolean",
    "Company": "Company",
    "CompanyFailureNotFound": "CompanyFailureNotFound",
    "CompanyFilterInput": "CompanyFilterInput",
    "CompanyFilterOperationInput": "CompanyFilterOperationInput",
    "CompanyInput": "CompanyInput",
    "CompanyOutput": "CompanyOutput",
    "CompanyQuery": "CompanyQuery",
    "CompanySuccess": "CompanySuccess",
    "CountryCode": "CountryCode",
    "Cursor": "Cursor",
    "Date": "Date",
    "DateFilter": "DateFilter",
    "DateTime": "DateTime",
    "EditCompanyFailureNotFound": "EditCompanyFailureNotFound",
    "EditCompanyInput": "EditCompanyInput",
    "EditCompanyOutput": "EditCompanyOutput",
    "EditCompanySuccess": "EditCompanySuccess",
    "EditPersonInput": "EditPersonInput",
    "EditPersonOutput": "EditPersonOutput",
    "EditPersonSuccess": "EditPersonSuccess",
    "Email": "Email",
    "FailureOutput": "FailureOutput",
    "FilterOperator": "FilterOperator",
    "Gender": "Gender",
    "Int": "Int",
    "InvalidCursorFailure": "InvalidCursorFailure",
    "LoginUserFailure": "LoginUserFailure",
    "LoginUserInput": "LoginUserInput",
    "LoginUserResponse": "LoginUserResponse",
    "LoginUserSuccess": "LoginUserSuccess",
    "Mutation": "Mutation",
    "NotFoundFailure": "NotFoundFailure",
    "NumberFact": "NumberFact",
    "NumberFactFailure": "NumberFactFailure",
    "NumberFactInput": "NumberFactInput",
    "NumberFactOutput": "NumberFactOutput",
    "NumberFactSuccess": "NumberFactSuccess",
    "PageInfo": "PageInfo",
    "PaginationInput": "PaginationInput",
    "Person": "Person",
    "PersonFilterInput": "PersonFilterInput",
    "PersonFilterOperationInput": "PersonFilterOperationInput",
    "PersonInput": "PersonInput",
    "PersonSortField": "PersonSortField",
    "PersonSortInput": "PersonSortInput",
    "PersonalIdentityCode": "PersonalIdentityCode",
    "PersonsOutput": "PersonsOutput",
    "PersonsPaginationEdge": "PersonsPaginationEdge",
    "PersonsPaginationResponse": "PersonsPaginationResponse",
    "Phone": "Phone",
    "Query": "Query",
    "RegisterFailure": "RegisterFailure",
    "RegisterFailureAlreadyExists": "RegisterFailureAlreadyExists",
    "RegisterInput": "RegisterInput",
    "RegisterResponse": "RegisterResponse",
    "RegisterSuccess": "RegisterSuccess",
    "RemoveEmployeeInput": "RemoveEmployeeInput",
    "RemoveEmployeeOutput": "RemoveEmployeeOutput",
    "RemoveEmployeeSuccess": "RemoveEmployeeSuccess",
    "SortOrder": "SortOrder",
    "String": "String",
    "StringFilter": "StringFilter",
    "Subscription": "Subscription",
    "TimeFilter": "TimeFilter",
    "Timestamp": "Timestamp",
    "UUID": "UUID",
    "Underage": "Underage",
    "UniqueConstraintViolationFailure": "UniqueConstraintViolationFailure",
    "Upload": "Upload",
    "User": "User",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
  },
  "astNode": Object {
    "kind": "SchemaDefinition",
    "operationTypes": Array [
      Object {
        "kind": "OperationTypeDefinition",
        "operation": "query",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 1401,
              "start": 1396,
            },
            "value": "Query",
          },
        },
      },
      Object {
        "kind": "OperationTypeDefinition",
        "operation": "mutation",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 371,
              "start": 363,
            },
            "value": "Mutation",
          },
        },
      },
      Object {
        "kind": "OperationTypeDefinition",
        "operation": "subscription",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "loc": Object {
              "end": 1640,
              "start": 1628,
            },
            "value": "Subscription",
          },
        },
      },
    ],
  },
  "description": undefined,
  "extensionASTNodes": Array [],
  "extensions": undefined,
}
`;

exports[`Graphql / schema createSchema 1`] = `
Object {
  "inheritResolversFromInterfaces": true,
  "resolverValidationOptions": Object {
    "requireResolversForAllFields": "ignore",
    "requireResolversForResolveType": "ignore",
    "requireResolversToMatchSchema": "warn",
  },
}
`;
